src/java/org/apache/cassandra/config/registry/ConfigurationRegistry.java

```
@@ -0,0 +1,401 @@
+        }
+        finally
+        {
+            rwLock.readLock().unlock();
+        }
+    }
+
+    @Override public String getString(String name)
+    {
+        return get(String.class, name);
+    }
+
+    /**
+     * @param name the property name to check.
+     * @return {@code true} if the property with the given name is available, {@code false} otherwise.
+     */
+    @Override public boolean contains(String name)
+    {
+        lazyInit();
+        return properties.containsKey(name);
+    }
+
+    /**
+     * Returns a set of all the property names.
+     * @return set of all the property names.
+     */
+    @Override public Iterable<String> keys()
+    {
+        lazyInit();
+        return properties.keySet();
+    }
+
+    /**
+     * @param name The property name to get the type for.
+     * @return Property type for the property with the given name.
+     */
+    @Override public Class<?> type(String name)
+    {
+        lazyInit();
+        validatePropertyExists(properties.get(name), name);
+        return properties.get(name).getType();
+    }
+
+    /**
+     * @return The number of properties.
+     */
+    @Override public int size()
+    {
+        lazyInit();
+        return properties.size();
+    }
+
+    /**
+     * @param name The property name to get the type for.
+     * @return Property type for the property with the given name.
+     */
+    public boolean isWritable(String name)
+    {
+        validatePropertyExists(properties.get(name), name);
+        return properties.get(name).isWritable();
+    }
+
+    /**
+     * Adds a listener for the property with the given name.
+     * @param name property name to listen to.
+     * @param listener listener to add.
+     * @param <T> type of the property.
+     */
+    public <T> void addPropertyChangeListener(String name, ConfigurationListener.ChangeType type, ConfigurationListener<T> listener, Class<T> listenerType)
```

if you listen to a property that is `@Replaced` what *should* happen?  The common case would be a bug (we forgot to update), but there *might* be a case where you want to see if this is happening (don't know of one)...

Maybe we should detect this and fail fast, force the caller to say "I know what I am doing, I want the old name!"?

---

In my view, if you change the name of a property, you must also update the corresponding listener usage. I propose to enforce this rule and avoid supporting different names for the same properties in internal code, so only limited subsystems will work with the `Replaced` annotation and know the old names.

The `fail fast` rule will work once the `addPropertyChangeListener` is fired with an incorrect property name.